core/src/main/java/software/amazon/awssdk/interceptor/ExecutionInterceptorChain.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.interceptor;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import software.amazon.awssdk.SdkRequest;
+import software.amazon.awssdk.SdkResponse;
+import software.amazon.awssdk.annotation.SdkInternalApi;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.interceptor.context.DefaultFailedExecutionInterceptorContext;
+import software.amazon.awssdk.interceptor.context.InterceptorContext;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A wrapper for a list of {@link ExecutionInterceptor}s that ensures the interceptors are executed in the correct order as it
+ * is documented in the {@link ExecutionInterceptor} documentation.
+ *
+ * Interceptors are invoked in forward order up to {@link #beforeTransmission} and in reverse order after (and including)
+ * {@link #afterTransmission}. This ensures the last interceptors to modify the request are the first interceptors to see the
+ * response.
+ */
+@SdkInternalApi
+public class ExecutionInterceptorChain {
+    private static final Logger LOG = Logger.loggerFor(ExecutionInterceptorChain.class);
+
+    private final List<ExecutionInterceptor> interceptors;
+
+    /**
+     * Create a chain that will execute the provided interceptors in the order they are provided.
+     */
+    public ExecutionInterceptorChain(List<ExecutionInterceptor> interceptors) {
+        this.interceptors = new ArrayList<>(Validate.paramNotNull(interceptors, "interceptors"));
+        LOG.debug(() -> "Creating an interceptor chain that will apply interceptors in the following order: " + interceptors);
+    }
+
+    public void beforeExecution(InterceptorContext context, ExecutionAttributes executionAttributes) {
+        interceptors.forEach(i -> i.beforeExecution(context, executionAttributes));
+    }
+
+    public InterceptorContext modifyRequest(InterceptorContext context, ExecutionAttributes executionAttributes) {
+        InterceptorContext result = context;
+        for (ExecutionInterceptor interceptor : interceptors) {
```

Comment:
I tried it out, but it was hard to read because the interface didn't fit perfectly (it's built to not assume serial execution of reduction operations, so you have to jump through hoops to make it work). I might have missed something that could have simplified it, though. I'm open to a proposed implementation!